rollbar-android/src/main/java/com/rollbar/android/notifier/sender/ConnectivityDetector.java

```diff
@@ -0,0 +1,182 @@
+      // We don't know, assume connection is available.
+      return true;
+    }
+
+    // We programmatically check for this permission in the constructor, and return early
+    // in this method if we don't have it.
+    @SuppressLint("MissingPermission")
+    NetworkInfo network = connectivityManager.getActiveNetworkInfo();
+    if (network == null) {
+      return false;
+    }
+
+    return network.isAvailable() && network.isConnected();
+  }
+
+  private boolean hasNetworkStatePermission() {
+    String permission = Manifest.permission.ACCESS_NETWORK_STATE;
+    int res = androidContext.checkPermission(permission,
+            android.os.Process.myPid(), android.os.Process.myUid());
+    return res == PackageManager.PERMISSION_GRANTED;
+  }
+
+  private void registerNetworkStateReceiver() {
+    synchronized (this.receiverLock) {
+      unregisterNetworkStateReceiver();
+      this.receiver = new NetworkUpBroadcastReceiver();
+    }
+
+    this.androidContext.registerReceiver(this.receiver,
+            new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
+
+    if (this.androidContext instanceof Activity) {
+      // Allows examples and documentation to remain unmodified even if users enable the
+      // connection state monitoring feature, otherwise all SDK users would need to add
+      // logic to their `onDestroy` methods to stop the IntentReceiver from leaking.
+      Activity targetActivity = (Activity) this.androidContext;
+      if (targetActivity.getApplication() != null) {
+        ReceiverActivityCallback activityCallback =
+                new ReceiverActivityCallback(targetActivity);
+        targetActivity.getApplication()
+                .registerActivityLifecycleCallbacks(activityCallback);
+      }
+    }
+  }
+
+  private void unregisterNetworkStateReceiver() {
+    synchronized (this.receiverLock) {
+      if (this.receiver != null && this.androidContext != null) {
+        this.androidContext.unregisterReceiver(this.receiver);
+        this.receiver = null;
+      }
+    }
+  }
+
+  private class NetworkUpBroadcastReceiver extends BroadcastReceiver {
+    @Override
+    public void onReceive(Context context, Intent intent) {
+      if (isNetworkAvailable()) {
+        Runnable signal = networkRestoredSignal;
+        if (signal != null) {
+          signal.run();
+        }
+      }
+      // We resume based on a connection event, but we don't suspend based on it. We err on
+      // the side of trying to send the occurrences even if we suspect the network is down,
+      // since connectivity detection involves heuristics that aren't 100% reliable (eg. some
+      // test endpoint such as http://clients3.google.com/generate_204 might be temporarily
+      // unavailable or blocked, but https://api.rollbar.com might be working.)
+      // We only suspend sending occurrences when a send attempt to the configured Rollbar
+      // endpoint fails due to a network availability issue.
```

Comment:
We do check that, by attempting to send an occurrence to `api.rollbar.com`, as soon as one is ready to be sent.

There's no use for us in knowing that the network is down if we don't have an occurrence that needs to be sent, and it would generate unnecessary network traffic.

The only question here is whether we should also postpone the check when we detect that the connection is back up, but `isNetworkAvailable()` doesn't performing any network checks itself, it piggybacks on checks done by Android's `ConnectivityManager`, which serves the OS and many other applications. So querying it ourselves is unlikely to cause any additional traffic, and keeps the code simple.

Reply:
The summary from the Slack conversation:
We follow the principle that the only valid reachability check of a remote endpoint, is to actually try to perform the operation we intend to perform. Ie, firewalls, routes, MITM packet inspection proxies, etc... can block all sorts of things we might use as a signal to figure out reachability, to possible extremes of blocking a "GET" request to Rollbar's  `status/ping` endpoint, but not a "POST" which is what we need to send an occurrence.

In terms of CPU usage, the queue is not locked while sending is suspended, the sender thread is free as well, and the item is out of the queue while we're figuring out if a failure is due to the network being down. This is all bit obscured by some small hoops I jumped through to keep everything backwards compatible, such as not changing the queue type, and getting the result of sending an occurrence via a listener instead of a return of `sender.send` (which currently returns void.)

Finally we suspend sending immediately after an occurrence that has failed because the network is unavailable (defined by the heuristics in `ConnectionAwareSenderFailureStrategy`), we don't send a full batch. So the first occurrence that fails means sending is suspended, no matter how large the queue is.

Reply:
Great summary! Marking it as resolved now.